pilot/pkg/model/extensions.go

```diff
@@ -0,0 +1,129 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package model
+
+import (
+	"net/url"
+	"strings"
+	"time"
+
+	envoy_config_core_v3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	envoy_extensions_filters_http_wasm_v3 "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/wasm/v3"
+	envoy_extensions_wasm_v3 "github.com/envoyproxy/go-control-plane/envoy/extensions/wasm/v3"
+	"github.com/gogo/protobuf/types"
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes/any"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/types/known/durationpb"
+
+	extensions "istio.io/api/extensions/v1alpha1"
+	"istio.io/istio/pkg/config"
+	"istio.io/istio/pkg/util/gogo"
+)
+
+const (
+	defaultRuntime = "envoy.wasm.runtime.v8"
+)
+
+type WasmPluginWrapper struct {
+	extensions.WasmPlugin
+
+	Name      string
+	Namespace string
+
+	ExtensionConfiguration *envoy_extensions_filters_http_wasm_v3.Wasm
+}
+
+func convertToWasmPluginWrapper(plugin *config.Config) *WasmPluginWrapper {
+	var ok bool
+	var wasmPlugin *extensions.WasmPlugin
+	if wasmPlugin, ok = plugin.Spec.(*extensions.WasmPlugin); !ok {
+		return nil
+	}
+
+	cfg := &any.Any{}
+	if wasmPlugin.PluginConfig != nil && len(wasmPlugin.PluginConfig.Fields) > 0 {
+		cfgJSON, err := protojson.Marshal(proto.MessageV2(wasmPlugin.PluginConfig))
+		if err != nil {
+			log.Warnf("wasmplugin %v/%v discarded due to json marshaling error: %s", plugin.Namespace, plugin.Name, err)
+			return nil
+		}
+		cfg = gogo.MessageToAny(&types.StringValue{
+			Value: string(cfgJSON),
+		})
+	}
+	var datasource *envoy_config_core_v3.AsyncDataSource
+	var sha256 string
+	u, err := url.Parse(wasmPlugin.Url)
+	if err != nil {
+		log.Warnf("wasmplugin %v/%v discarded due to failure to parse URL: %s", plugin.Namespace, plugin.Name, err)
+		return nil
+	}
+	if wasmPlugin.XSha256 == nil {
```

Comment:
I  suspect it will fix the unmarshalling issues and `"nil"` workaround as well

Reply:
agreed; this has caused me nothing but pain :slightly_smiling_face: I'll do this after this PR has merged though, because I don't want to delay it even further by waiting on an additional API PR to merge

Reply:
Ok, lets make sure we do it before we ship this or it will be a breaking change I think